### PR TITLE
const views

### LINKED ISSUE #56: const tensor modification through views
ASGarD
    asgardasgard
    Issues
    #56

Open
Opened 2 months ago by Graham Lopez
Close issue
New issue
const tensor modification through views

a const owner can be modified through a view

possibly make view ctors private, and only create through public owner met
Related issues
0

Graham Lopez @bgl added bug label 2 months ago

### PR DESCRIPTION
**Pull request by bmcdanie**
closes #56

### CODE REVIEW

**Comment by grahamlopez** [comment]
test this please

File: src/tensors.hpp [hunk 1]
```diff
@@ -375,27 +457,26 @@ class matrix
...
   template<mem_type omem, resource r_ = resrc, typename = enable_for_host<r_>>
   matrix<P> kron(matrix<P, omem> const &) const;
 
   // clang-format off
-  template<typename U=P, resource r_ = resrc,
-	  typename = enable_for_host<r_>>
-  std::enable_if_t<
-    std::is_floating_point<U>::value && std::is_same<P, U>::value,
+  template<typename U = P, mem_type m_ = mem,
+           typename = disable_for_immutable<m_>, resource r_ = resrc,
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/tensors.hpp]: done

File: src/tensors.hpp [hunk 2]
```diff
@@ -23,17 +23,36 @@ class access_badge
...
+    std::enable_if_t<mem == mem_type::view || mem == mem_type::const_view>;
+
+// enable only for const views
+template<mem_type mem>
+using enable_for_const_view = std::enable_if_t<mem == mem_type::const_view>;
+
+// enable only for mutable views
 template<mem_type mem>
-using enable_for_view = std::enable_if_t<mem == mem_type::view>;
+using enable_for_mutable_view = std::enable_if_t<mem == mem_type::view>;
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/tensors.hpp]: done

File: src/tensors.hpp [hunk 3]
```diff
@@ -696,36 +806,40 @@ fk::vector<P, mem, resrc>::vector(fk::matrix<P> const &mat)
...
+{}
 
 // delegating constructor to extract view from owner. overload for default case
 // of viewing the entire owner
+// const view version
 template<typename P, mem_type mem, resource resrc>
 template<mem_type, typename>
 fk::vector<P, mem, resrc>::vector(
     fk::vector<P, mem_type::owner, resrc> const &a)
     : vector(a, 0, std::max(0, a.size() - 1))
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/tensors.hpp]: done

**Pull request merged**